_posts/2021-03-29-apply-pretrained-pytorch-model.md

```diff
@@ -0,0 +1,153 @@
+2. Load a pre-trained PyTorch model that featurizes images
+3. Construct a function to apply the model onto each chunk
+4. Apply that function across the Dask array with the dask.array.map_blocks function.
+5. Store the result back into Zarr format
+
+### Step 1. Load the image data
+
+First, we load the image data into a Dask array.
+
+ We were provided sample data from [Prof. Shroff’s lab](https://www.nibib.nih.gov/about-nibib/staff/hari-shroff) at the NIH. The data originally was provided as a 3D TIFF file which we subsequently converted to Zarr with a shape of (950, 2048, 2048). That's a z-stack of 950 images where each slice is 2048x2048 pixels.
+
+This is the same data that we analysed in our last [blogpost on Dask and ITK](https://blog.dask.org/2019/08/09/image-itk). You should note the similarities to that workflow even though we are now using new libaries and performing different analyses.
+
+```
+cd '/Users/nicholassofroniew/Github/image-demos/data/LLSM'
+```
+
+```python
+# Load our data
+import dask.array as da
+imgs = da.from_zarr("AOLLSM_m4_560nm.zarr")
+imgs
+```
+
+```
+dask.array<from-zarr, shape=(20, 199, 768, 1024), dtype=float32, chunksize=(1, 1, 768, 1024)>
+```
+
+### Step 2. Load a pre-trained PyTorch model
+
+Next, we load our pre-trained UNet model.
+
+This UNet model takes in an 2D image and returns a 2D x 16 array, where each pixel is now associate with a feature vector of length 16.
+
+We thank Mars Huang for training this particular UNet on a corpous of biological images to produce biologically relevant feature vectors, as part of his work on [interactive bio-image segmentation](https://github.com/transformify-plugins/segmentify). These features can then be used for more downstream image processing tasks such as image segmentation.
+
+
+```python
+# Load our pretrained UNet¶
+import torch
+from segmentify.model import UNet, layers
+
+def load_unet(path):
+    """Load a pretrained UNet model."""
+
+    # load in saved model
+    pth = torch.load(path)
+    model_args = pth['model_args']
+    model_state = pth['model_state']
+    model = UNet(**model_args)
+    model.load_state_dict(model_state)
+
+    # remove last layer and activation
+    model.segment = layers.Identity()
+    model.activate = layers.Identity()
+    model.eval()
+
+    return model
+
+model = load_unet("HPA_3.pth")
+```
+
+### Step 3. Construct a function to apply the model to each chunk
+
+We make a function to apply our pre-trained UNet model to each chunk of the Dask array.
+
+Because Dask arrays are just made out of Numpy arrays which are easily converted to Torch arrays, we're able to leverage the power of machine learning at scale.
+
+```python
+# Apply UNet featurization
+import numpy as np
+
+def unet_featurize(image, model):
+    """Featurize pixels in an image using pretrained UNet model.
+    """
+    import numpy as np
+    import torch
+
+    # remove leading two length-one dimensions
```

Comment:
Should we add a note here?

From the 
Matt Rocklin writes:
> These lines are interesting:
> ```python
>     # remove leading two length-one dimensions
>     img = image[0, 0, ...]
>    
>     # make sure image has four dimentions (b,c,w,h)
>     img = np.expand_dims(np.expand_dims(img, 0), 0)
>     img = np.transpose(img, (1,0,2,3))
> ```
> You take away two dimensions, and then seem to add them back in. Was this just because of copy-pasting from the previous post, or was this intentional in some way.
>
> Is there a nice way to produce an image from the output data here? One of the things that I think we could have done better on with the deconvolution post was providing a before-and-after comparison. Presumably pixel-level segmentation would also provide some eye-candy.

Nick answers:
> I wasn't sure what to do with those two lines, removing then adding back in the dimensions. On the one hand they accomplish nothing, but in many ways that is by chance and the reasons they are in there (if explained!) might be informative to people.
> 
> My understanding is that the first two lines where we drop the leading two dimensions are because of the shape of the original dask array and the chunking - we have a 4D array but we're interested in extracting 2D images so we drop those first two length-one dimensions. If we'd had a 5D array of data we would have had to drop the first three length-one dimensions.
>
> The adding of dimensions back in corresponds to getting the data in a shape that PyTorch expects, which is Batch x Channel x Width x Height and here Batch and Channel are just 1, so in this example we go 4D -> 2D -> 4D, but we might have had to go 5D -> 2D -> 4D.
>
> Maybe all that is just more confusing than necessary and we can leave the comment # make sure image has four dimentions (b,c,w,h) without actually running that code. Also that transpose doesn't seem to do anything either in this case (which was a copy and paste type of situation!) and so can be dropped too.
> 
> As to producing images of the output data, yes it's possible - the output data can actually be used to produce 16 images of the different features, but they don't look so great. Maybe I'll think of something else to show. I agree though a before / after image would be nice.